### PR TITLE
feat: Startup logic

### PR DESCRIPTION
**Pull request by LiamMartens**
@six7 I streamlined the startup logic as I had mentioned before. Now the plugin data will be pre-fetched in one go and sent to the front-end where it will be handled in a single step-by-step process reducing the amount of round-trips and making the plugin logic more clear to reduce the possibility for errors.

I did alter the loading behavior of the plugin slightly where the initial pulling of the tokens will now also happen during the loading screen, but the loading screen has also been updated with more clear information for the user (clearly stating the different steps; such as verifying the license, getting the LD flags etc, ...).

The process now also ensures the LD flags are ready before initializing any other logic in the plugin ensuring the right features are available to the user at the right time.

As a TODO to myself - I need to adjust the tests where required

### CODE REVIEW

**Comment by LiamMartens** [comment]
https://user-images.githubusercontent.com/5265324/180692517-f7a59109-eedd-4ff7-a068-65ee9704abc8.mp4

**Comment by six7** [comment]
This is awesome, I love the various loading states! I tried with a couple different scenarios but will do some more testing and then look at the code. Thank you for this! 🎉

**Comment by LiamMartens** [comment]
@six7 updated all the tests etc..

File: src/app/components/AppContainer/__tests__/AppContainerIntegration.test.tsx [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    identify: () => Promise.resolve(),
+  }),
+  useFlags: () => ({}),
+}));
+
+describe('AppContainer (integration)', () => {
+  beforeEach(resetSuite);
+
+  beforeAll(() => {
+    // useLDClientSpy.mockImplementation(() => ({
```
Comment (six7) [inline src/app/components/AppContainer/__tests__/AppContainerIntegration.test.tsx:194]: is this supposed to commented out?

File: src/app/components/AppContainer/__tests__/AppContainerIntegration.test.tsx [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    //   // identify: () => Promise.resolve(),
+    // } as any));
+    // useFlagsSpy.mockImplementation(() => ({
+    // }));
+  });
+
+  afterAll(() => {
+    resetSuite();
+    // useFlagsSpy.mockReset();
```
Comment (six7) [inline src/app/components/AppContainer/__tests__/AppContainerIntegration.test.tsx:203]: same, do we need those comments?

File: src/app/components/AppContainer/startupProcessSteps/saveStorageInforamationFactory.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import useStorage from '@/app/store/useStorage';
+import type { Dispatch } from '@/app/store';
+import type { StartupMessage } from '@/types/AsyncMessages';
+
+export function saveStorageInforamationFactory(
```
Comment (six7) [inline src/app/components/AppContainer/startupProcessSteps/saveStorageInforamationFactory.ts:5]:
```suggestion
export function saveStorageInformationFactory(
```

should probably change any occurrence of this typo

File: src/app/components/LaunchDarkly/LDProvider.tsx [hunk 4]
```diff
@@ -12,8 +11,7 @@ const ldClientSideId = process.env.LAUNCHDARKLY_SDK_CLIENT || '';
 
 export const LDProviderWrapper = ({ children }: LDProviderProps) => {
   const userId = useSelector(userIdSelector);
-  const licenseKey = useSelector(licenseKeySelector);
-  if (userId && licenseKey) {
+  if (userId) {
```
Comment (six7) [inline src/app/components/LaunchDarkly/LDProvider.tsx:14]: just to confirm: was removing license key here intentional? i remember we limited this to just call launchdarkly when a license key was present to reduce the amount of API calls we need (as those have a plan limit)

File: src/utils/uiSettings.ts [hunk 5]
```diff
@@ -62,9 +62,12 @@ export async function getUISettings(): Promise<SavedSettings> {
         inspectDeep,
       };
 
-      notifyUISettings(settings);
+      if (notify) {
+        notifyUISettings(settings);
+      }
     }
   } catch (err) {
+    console.error(err);
```
Comment (six7) [inline src/utils/uiSettings.ts:70]: just so i understand: when should we use `handleReactError` and when just a console err? It feels like we'd always want `handleReactError` as we can then track these?

**Review by six7 — CHANGES_REQUESTED** [review]
haven't fully functionally tested yet but went through code and had just a few minor things to note. this is great, way more understandable!

**Comment by six7** [comment]
One thing I noticed: If there's a credentials error during startup the plugin stays in `Fetching credentials` loading state forever. This happened with ADO sync.

image

File: src/utils/uiSettings.ts (hunk 5 above)
Comment (LiamMartens) [reply src/utils/uiSettings.ts:70]: The `uiSettings.ts` file is used on the plugin side - not the react side; we can't really track much on the plugin side due to the sandboxing - so `handleReactError` should only be used in the UI side

File: src/app/components/LaunchDarkly/LDProvider.tsx (hunk 4 above)
Comment (LiamMartens) [reply src/app/components/LaunchDarkly/LDProvider.tsx:14]: Ahh I see - I'll re-add it and add a comment in code

**Comment by LiamMartens** [comment]
@six7 the error occurs if the local api credentials are invalid or something?

**Comment by LiamMartens** [comment]
@six7 issue with the loading state should be fixed

**Review by six7 — APPROVED** [review]

**Comment by six7** [comment]
Can you resolve merge conflicts? After that let's merge this in.

**Pull request merged**